tf_quant_finance/math/jacobian.py

```diff
@@ -0,0 +1,120 @@
+# limitations under the License.
+"""Helper functions for computing jacobian."""
+
+
+import functools
+
+import tensorflow.compat.v2 as tf
+
+
+def jacobian(func, xs, unconnected_gradients=None, name=None,
+             parallel_iterations=None, experimental_use_pfor=True):
+  """Computes the jacobian of `func` wrt to `xs`.
+  Args:
+    func: Python callable accepting one `Tensor` of shape of `x` and returning
+      a `Tensor` of any shape. The function whose jacobian is to be computed.
+    xs: A `Tensor` of parameters of `f` for which to differentiate. The
+      dimension (ndim) should be 2, and the first dimension is the batch
+      shape. If the dimension is 1, the `Tensor` of parameters is converted
+      into a single batched parameter with the dimension 2.
+    unconnected_gradients: An enum `tf.UnconnectedGradients` which specifies the
+      gradient value returned when the given input tensors are unconnected.
+      Default value: `None`, which maps to `tf.UnconnectedGradients.NONE`.
+    name: Python `str` name prefixed to ops created by this function.
+      Default value: `None` (i.e., 'jacobian').
+    parallel_iterations: A knob to control how many iterations are dispatched
+      in parallel. This knob can be used to control the total memory usage.
+    experimental_use_pfor: If true, uses pfor for computing the Jacobian.
+      Else uses a tf.while_loop.
+  Returns:
+    A `Tensor` with the gradient of `y` wrt each of `x`.
+  """
+  unconnected_gradients = unconnected_gradients or tf.UnconnectedGradients.NONE
+  xs, is_xs_batch_size = _prepare_args(xs)
+  with tf.name_scope(name or "jacobian"):
+    if not callable(func):
+      raise ValueError("`func_or_y` should be a callable in eager mode or "
+                       "when `tf.GradientTape` is used.")
+    with tf.GradientTape() as tape:
+      tape.watch(xs)
+      y = func(xs)
+    jac = tape.batch_jacobian(
+        y, xs,
+        unconnected_gradients=unconnected_gradients,
+        parallel_iterations=parallel_iterations,
+        experimental_use_pfor=experimental_use_pfor
+    )
+
+    if is_xs_batch_size:
+        return jac
+
+    return jac[0]
+
+
+def value_and_jacobian(f, xs, unconnected_gradients=None, name=None,
+                       parallel_iterations=None, experimental_use_pfor=True):
+  """Computes `f(xs)` and its jacobian wrt to `xs`.
+  Args:
+    f: Python `callable` to be differentiated. If `f` returns a scalar, this
+      scalar will be differentiated. If `f` returns a tensor or list of tensors,
+      by default a scalar will be computed by adding all their values to produce
+      a single scalar. If desired, the tensors can be elementwise multiplied by
+      the tensors passed as the `dy` keyword argument to the returned jacobian
+      function.
+    xs: A `Tensor` of parameters of `f` for which to differentiate. The
+      dimension (ndim) should be 2, and the first dimension is the batch
```

Comment:
same as above

Reply:
As discussed before, addressed in 8a357ce